printer/doc.go

```diff
@@ -0,0 +1,94 @@
+/*
+Printer provides features for printing out IPLD nodes and their contained data in a human-readable diagnostic format.
+
+Outputs should look like...
+
+	map{
+		"foo": "bar"
+		"zot": struct<Foobar>{
+			someFieldName: list{
+				0: "this"
+				1: "is untyped"
+				2: int{400}
+				3: bool{true}
+			}
+			otherField: list<ANamedListType>{
+				0: "mind you: 'ANamedListType' is the name of the *list type*."
+				1: "it is not the name of the value types.  those are not actually shown here."
+				2: "you'd have to look at the schema for that information."
+			}
+			moreField: list<[nullable String]>{
+				0: "this is a typed list"
+				1: "but anonymous"
+				2: null
+			}
+		}
+		"frog": map<{String:String}>{
+			"as you have probably imagined": "this is a typed (but anonymous type) map"
+		}
+		"numbers": int{1}
+		"binary": bytes{ABCDEF0123456789}
+		"typed numbers": int<MyNamedTypeInt>{9000}
+		"typed string": string<MyNamedTypeString>{"okay, this one needed some marker prefixes."}
+		"map with typed keys": map<{MyNamedTypeString:MyNamedTypeString}>{
+			"despite being typed": string<MyNamedTypeString>{"map keys still never need prefixes; there's no ambiguity"}
+			"but for the values": string<MyNamedTypeString>{"we still use them"}
+			"well, maybe": string<MyNamedTypeString>{"this might actually be worth debating; it doesn't seem necessary (unless the map value is a variant type, but that's then already addressed by another case)"}
+		}
+		"structs": struct<FooBar>{
+			foo: "do not need to have quoted field names"
+			bar: "because (unlike map keys) their character range is already restricted"
+		}
+		"unit types": unit<TheTypeName>
+		"notice unit types": "have no braces at all, because they have literally no further details.  they're all type info."
+		"variants": variant<TheUnionName>{string<TheInhabitant>{
```

Comment:
so `variant` is the name for "specific variant of a union"? introduction of new terminology is a concern I reckon, it doesn't read as smoothly if there's disconnect with user-consumed schema lingo

Reply:
oh oof, I should `s/variant/union/` this.  It's a stretchmark from how I think during development: I was trying out "variant" as an alternate name for "union" and trying to get a feel for if I liked it when seeing and using it for a while.

The result is "yes", I do kinda come to prefer that term.  I think maybe we should consider doing a mass rename of "union" to "variant".

But that's not a discussion I want to force through in this PR; it deserves its own space.